presto-main/src/main/java/com/facebook/presto/server/security/PrestoAuthenticator.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server.security;
+
+import com.facebook.airlift.http.server.AuthenticationException;
+import com.facebook.airlift.http.server.Authenticator;
+
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+
+import java.security.Principal;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrestoAuthenticator
+        implements Authenticator
+{
+    private PrestoAuthenticatorManager authenticatorManager;
+
+    @Inject
+    public PrestoAuthenticator(PrestoAuthenticatorManager authenticatorManager)
+    {
+        this.authenticatorManager = requireNonNull(authenticatorManager, "authenticatorManager is null");
+        authenticatorManager.setRequired();
+    }
+
+    @Override
+    public Principal authenticate(HttpServletRequest request)
+            throws AuthenticationException
+    {
+        try {
+            return authenticatorManager.getAuthenticator().createAuthenticatedPrincipal(request);
```

Comment:
`HttpServletRequest` is a little broad because it also returns the input stream of the actual request.  In theory, someone could call that in a plugin.  I think that should be a loud failure if that happens.  Can you implement a subclass of `HttpServletRequest`, which wraps an inner `HttpServletRequest`, but throws if you attempt to use any method which modifies the request, or reads something which doesn't support repeatable reads (such as `getInputStream`)?

Reply:
I've implemented the changes as suggested. I created a subclass of ```HttpServletRequest``` that wraps an inner ```HttpServletRequest``` and throws an exception if any method is called that modifies the request or accesses non-repeatable data like getInputStream. Let me know if there's anything else you'd like me to adjust.